### PR TITLE
Local Sync Agent

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
We want to support synchronous tasks in the agent.

## What changes were proposed in this pull request?
Add sync agent local execution mode.

## How was this patch tested?
I test this patch in this [chatgpt agent](https://github.com/flyteorg/flytekit/pull/2086) PR.

### Screenshots
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/1822
https://github.com/flyteorg/flytekit/pull/2086

### CODE REVIEW

File: flytekit/extend/backend/base_agent.py [hunk 1]
```diff
@@ -158,15 +159,27 @@ def get_agent_secret(secret_key: str) -> str:
...
+    grpc_ctx = MagicMock(spec=grpc.ServicerContext)
+    return grpc_ctx
+
+
 class AsyncAgentExecutorMixin:
     """
     This mixin class is used to run the agent task locally, and it's only used for local execution.
     Task should inherit from this class if the task can be run in the agent.
+    It can handle asynchronous tasks and synchronous tasks.
+    Asynchronous tasks are for tasks running long, for example, running query job.
```
Comment (pingsutw) [inline flytekit/extend/backend/base_agent.py:174]:
```suggestion
    Asynchronous tasks are tasks that take a long time to complete, such as running a query.
```

File: flytekit/extend/backend/base_agent.py [hunk 2]
```diff
@@ -158,15 +159,27 @@ def get_agent_secret(secret_key: str) -> str:
...
+    Synchronous tasks are for tasks running quick, for example, you want to execute something really fast, or even retrieving some metadata from a backend service.
```
Comment (pingsutw) [inline flytekit/extend/backend/base_agent.py:175]:
```suggestion
    Synchronous tasks run quickly and can return their results instantly. Sending a prompt to ChatGPT and getting a response, or retrieving some metadata from a backend system.
```

File: flytekit/extend/backend/base_agent.py [hunk 3]
```diff
@@ -198,34 +218,30 @@ async def _create(
...
 
         # Convert python inputs to literals
         literals = inputs or {}
         for k, v in inputs.items():
             literals[k] = TypeEngine.to_literal(ctx, v, type(v), self._entity.interface.inputs[k].type)
-
-        literal_map = LiteralMap(literals)
-
-        if isinstance(self, PythonFunctionTask):
+        literal_map = LiteralMap(literals) if literals else None
```
Comment (pingsutw) [inline flytekit/extend/backend/base_agent.py:226]:
```suggestion
        literal_map = LiteralMap(literals)
```

File: flytekit/extend/backend/base_agent.py [hunk 4]
```diff
@@ -198,34 +218,30 @@ async def _create(
...
+        if literal_map and isinstance(self, PythonFunctionTask):
```
Comment (pingsutw) [inline flytekit/extend/backend/base_agent.py:227]: `pyflyte-execute` has some issues, so we should always upload `literals.pb` even if the literal is None.
check out https://github.com/flyteorg/flytekit/pull/2054

Comment (Future-Outlier) [reply flytekit/extend/backend/base_agent.py:227]: Thank you!

**Review by pingsutw — APPROVED** [review]

**Pull request merged**